### PR TITLE
Lean Optimization interface in QCAlgorithm

### LINKED ISSUE #578: Create IOptimizer Interface in QCAlgorithm
Allow users to set the optimization technique to use. Set the IOptimizer which will generate the parameter sets to backtest. 

IOptimizer would take in the last set of results; and return a new parameter set to backtest over.

E.g. A "GridSearchOptimization" would be called once with no results at the start; then it would return linear steps across each of the parameter variations to make up all possible combinations to backtest. A GenericOptimizer would remember breed its parameter sets as to get the best results and generate new ones. Bisect optimizer could bisect and skip large empty parts

The optimizer would be stateful; so could record result sets to perform searches over all results before returning the new parameter set.

### PR DESCRIPTION
**Pull request by AdalyatNazirov**
#### Description

Implement optimization tool to find the best values of parameters on `QCAclgorithm`.
- The core is `LeanOptimizer` that manage strategy and Lean executions. For local runs we provide `ConsoleLeanOptimizer`.  
- Strategies - we support [Grid Search strategy ](https://en.wikipedia.org/wiki/Brute-force_search)and Euler-alike strategy (we can't implement Euler Search because we don't have a mathematical function describing the algorithm). 
- Parameters - in current release we support Step based and arrays parameters. 
`OptimizationStepParameter` requires min and max values (boundaries) and step; 
`OptimizationArrayParameter` defining discrete values.
- Target - the parameter of Lean compute output is going to be maximized/minimized. Single. It's also possible to specify `TargetValue` meaning that optimization will stop once we reach the value. `Target` is any property in the backtest result json and should consist of target itself and all path, for example `Statistics.Sharpe Ratio`.
- Constraint - allows specifying statistical constraints for the optimization. Multiple (array). Solution and ParameterSet is not allowed if any of the constraint is not met.
 
#### Related Issue

Closes #578 

#### Motivation and Context

Allow users to set the optimization technique which generate the parameter sets to backtest and find the best values of parameters for specific algorithm. Currently provided Grid Search(aka Brute Force) and Euler-like algorithm to find the parameter values.

#### Requires Documentation Change

N/A

#### How Has This Been Tested?

- unit tests
- manual testing

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

**Comment by IlshatGaripov** [comment]
@AdalyatNazirov  i was interested by commit by involvement of mathematical calculations. What do we mean here by we support Grid Search strategy and Euler-alike strategy - I am wondeeinrg how is Euler search is used in solving optimization problem ?  I think we can refer to Ore's grapth theory to learn the fundamentalas and solve it programmatically.. - the basics for learning grapth theory https://www.amazon.com/THEORY-GRAPHS-Oystein-Ore/dp/B0032O6F8W  - yet not familiar with it

**Comment by AdalyatNazirov** [comment]
@IlshatGaripov Grid Search and Euler-like Search are the ways to generate set of parameters for Lean Algorithm to be run with. For example, we have parameters a = [1..100], b=[10..1000] (can be as many as you need). That strategies generate pairs of parameters (a,b), i.e. (1, 10), (2, 10) etc. After that we run Lean process for each parameters set and get Lean statistics for each parameter set. As a result of optimization process you find the best values of your parameters (a = 16, b = 524).

Grid Search = iterating range with the fixed step size ([https://en.wikipedia.org/wiki/Brute-force_search](https://en.wikipedia.org/wiki/Brute-force_search)). This method is good enough but it is not very accurate. On the one hand you need as small step as possible (to get accurate results), on the other hand you each Lean run takes some time and is not very cheap in term of performance. That's why we also developer another strategy with Euler method idea in mind  (it's not exactly Euler method for solving differential equations) - instead of iterating all data points on the grid with small step, we iterate grid with big step, and then reduce it on the next step etc etc. 
 
As you noticed there can be implemented many different strategies (I didn't investigate Ore's grapth theory though), but we considered [genetic algorithms](https://en.wikipedia.org/wiki/Genetic_algorithm), [heuristic technique](https://en.wikipedia.org/wiki/A*_search_algorithm), Monte-Carlo methods ([Simulated annealing](https://en.wikipedia.org/wiki/Simulated_annealing)).  In the end we decided to come up with something simple and straightforward in first revision. I believe we will implement more different strategies in future. For this case provide an interface `IOptimisationStrategy` to abstract specific optimization strategy.

**Comment by IlshatGaripov** [comment]
Ah ok thank you for explanation @AdalyatNazirov this makes the things clear, I thought it is Euler search from graph theory - I believe this is an algo : https://www.geeksforgeeks.org/eulerian-path-and-circuit/ - and wondered how this could be applied in solving the optimization problem - but it was different euler serach =) thanks,

**Review by Martin-Molinero — APPROVED** [review]

**Review by jaredbroad — COMMENTED** [review]
Minor comments here and notes from call to remove enumeration from parameters to encourage separation of strategy and parameter type concerns.

File: Optimizer/Objectives/Extremum.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+            _comparer = comparer;
+        }
+
+        /// <summary>
+        /// Compares two values; identifies whether condition is met or not.
+        /// </summary>
+        /// <param name="current">Left operand</param>
+        /// <param name="candidate">Right operand</param>
+        /// <returns>Returns the result of comparer with this arguments</returns>
+        public bool Better(decimal current, decimal candidate) => _comparer(current, candidate);
```
[the remark this answers is not in the document]
Comment (AdalyatNazirov) [reply Optimizer/Objectives/Extremum.cs:45]: yeah, that's why I used Better. Better is Better, and it doesn't matter smaller or larger :) do you have alternative name?

**Review by jaredbroad — APPROVED** [review]

**Comment by dakr001** [comment]
Hi, how far along is this project? Is it possible to run parameter search now, even just to test it out? Thanks.

**Pull request merged**